Expose algebraic floating point intrinsics

calder opened the pull request:
# Problem

A stable Rust implementation of a simple dot product is 8x slower than C++ on modern x86-64 CPUs. The root cause is an inability to let the compiler reorder floating point operations for better vectorization.

See https://github.com/calder/dot-bench for benchmarks. Measurements below were performed on a i7-10875H.

### C++: 10us ✅

With Clang 18.1.3 and `-O2 -march=haswell`:
<table>
<tr>
    <th>C++</th>
    <th>Assembly</th>
</tr>
<tr>
<td>
<pre lang="cc">
float dot(float *a, float *b, size_t len) {
    #pragma clang fp reassociate(on)
    float sum = 0.0;
    for (size_t i = 0; i < len; ++i) {
        sum += a[i] * b[i];
    }
    return sum;
}
</pre>
</td>
<td>
<img src="https://github.com/user-attachments/assets/739573c0-380a-4d84-9fd9-141343ce7e68" />
</td>
</tr>
</table>

### Nightly Rust: 10us ✅

With rustc 1.86.0-nightly (8239a37f9) and `-C opt-level=3 -C target-feature=+avx2,+fma`:
<table>
<tr>
    <th>Rust</th>
    <th>Assembly</th>
</tr>
<tr>
<td>
<pre lang="rust">
fn dot(a: &[f32], b: &[f32]) -> f32 {
    let mut sum = 0.0;
    for i in 0..a.len() {
        sum = fadd_algebraic(sum, fmul_algebraic(a[i], b[i]));
    }
    sum
}
</pre>
</td>
<td>
<img src="https://github.com/user-attachments/assets/9dcf953a-2cd7-42f3-bc34-7117de4c5fb9" />
</td>
</tr>
</table>

### Stable Rust: 84us ❌

With rustc 1.84.1 (e71f9a9a9) and `-C opt-level=3 -C target-feature=+avx2,+fma`:
<table>
<tr>
    <th>Rust</th>
    <th>Assembly</th>
</tr>
<tr>
<td>
<pre lang="rust">
fn dot(a: &[f32], b: &[f32]) -> f32 {
    let mut sum = 0.0;
    for i in 0..a.len() {
        sum += a[i] * b[i];
    }
    sum
}
</pre>
</td>
<td>
<img src="https://github.com/user-attachments/assets/936a1f7e-33e4-4ff8-a732-c3cdfe068dca" />
</td>
</tr>
</table>

# Proposed Change

Add `core::intrinsics::f*_algebraic` wrappers to `f16`, `f32`, `f64`, and `f128` gated on a new `float_algebraic` feature.

# Alternatives Considered

https://github.com/rust-lang/rust/issues/21690 has a lot of good discussion of various options for supporting fast math in Rust, but is still open a decade later because any choice that opts in more than individual operations is ultimately contrary to Rust's design principles.

In the mean time, processors have evolved and we're leaving major performance on the table by not supporting vectorization. We shouldn't make users choose between an unstable compiler and an 8x performance hit.

# References

* https://github.com/rust-lang/rust/issues/21690
* https://github.com/rust-lang/libs-team/issues/532
* https://github.com/rust-lang/rust/issues/136469
* https://github.com/calder/dot-bench
* https://www.felixcloutier.com/x86/vfmadd132ps:vfmadd213ps:vfmadd231ps

try-job: x86_64-gnu-nopt
try-job: x86_64-gnu-aux

saethlin commented on library/std/src/f32.rs:
```diff
@@ -1154,4 +1154,54 @@ impl f32 {
+    #[rustc_allow_incoherent_impl]
+    #[doc(alias = "fadd_algebraic", alias = "addAlgebraic")]
+    #[must_use = "method returns a new number and does not mutate the original value"]
+    #[unstable(feature = "float_algebraic", issue = "21690")]
```
The issue number here should be a tracking issue (i.e. based on the tracking issue template) which is specific to the feature. #21690 is not a tracking issue.

calder replied:
Filed https://github.com/rust-lang/rust/issues/136469 and updated links

saethlin commented on library/std/src/f32.rs:
```diff
@@ -1154,4 +1154,54 @@ impl f32 {
     }
+
+    /// Float addition that allows optimizations based on algebraic rules.
+    #[rustc_allow_incoherent_impl]
```
These methods should be in `core` so that they don't need this, and are also usable in `no_std` programs.

calder replied:
Updated, thanks!

calder commented:
Thanks for the speedy review @saethlin! I had a couple questions:
* Should we add these methods to `f16` and `f128` as well?
* Are intrinsic wrappers tested explicitly? I see tests for the intrinsics themselves but not for other wrappers.

RalfJung commented:
Thanks for the PR!

Going through the usual process, the first step would be a t-libs-api ACP to gauge the team's thinking on whether and how this should be exposed. This PR cannot land before a corresponding FCP has been accepted.

RalfJung commented:
In terms of tests, usually we have doc tests. Also, given that the semantics are far from obvious (these operations are nondeterministic!), they need to be documented more carefully - probably in some central location, which is then referenced from everywhere.

calder commented:
@RalfJung: Thanks for the quick response!
* Created ACP here: https://github.com/rust-lang/libs-team/issues/532
* Do you have a suggestion for a good central place to document these methods?

RalfJung commented:
https://doc.rust-lang.org/nightly/std/primitive.f32.html seems like a good place, that's where we already document everything special around NaNs. So a new section on algebraic operations there would probably be a good fit.

calder commented:
# Changes

* Added `f16` and `f128` methods as requested by @tgross35 [here](https://github.com/rust-lang/libs-team/issues/532#issuecomment-2630321075).
* Renamed to `algebraic_*()` to match `checked_*()` as requested by @tgross35 [here](https://github.com/rust-lang/libs-team/issues/532#issuecomment-2631224978).
* Added central documentation with verbage suggested by @Ralfjung [here](https://github.com/rust-lang/libs-team/issues/532#issuecomment-2630558384).

tgross35 commented on library/core/src/intrinsics/mod.rs:
```diff
@@ -2995,7 +2995,7 @@ pub unsafe fn float_to_int_unchecked<Float: Copy, Int: Copy>(_value: Float) -> I
 /// Float addition that allows optimizations based on algebraic rules.
 ///
-/// This intrinsic does not have a stable counterpart.
+/// Stabilized as [`f16::algebraic_add`], [`f32::algebraic_add`], [`f64::algebraic_add`] and [`f128::algebraic_add`].
```
Nit: the other docs have a consistent style for the message, this should probably also start with "The stable counterparts of this intrinsic are ...".

(Or "the public counterparts" like `select_unpredictable` since this isn't technically stable, but I think calling it "stable" is fine so we don't need to update it in the future).

calder replied:
I copied `float_to_int_unchecked` directly above it since
```rust
/// Stabilized as ...
```
is nice and concise compared to
```rust
/// The stable counterparts of this intrinsic are ...
```

Should we either leave these as is or update `float_to_unchecked` to the more verbose wording too?

tgross35 replied:
Ah so I guess there are a few others, no need to update anything that is already there.

tgross35 commented on library/core/src/primitive_docs.rs:
```diff
@@ -1317,6 +1317,15 @@ mod prim_f16 {}
+/// # Algebraic operators
+///
+/// Algebraic operators of the form `a.algebraic_*(b)` allow the compiler to reassociate individual
+/// floating point operations for better vectorization. Because of the unpredictable nature of
```
It would be good to give the user an idea of what to expect. Maybe between these two sentences add something like:

> The exact set of optimizations is unspecified but typically allows combining operations, rearranging series of operations based on mathematical properties, converting between division and reciprocal multiplication, and disregarding the sign of zero. This may mean that the results of elementary operations have undefined precision, but will not invoke input-dependent undefined behavior in performing these operations.

And then move "Because ..." to a new paragraph.

calder replied:
Added minus redundant "in performing these operations". How's that look?

tgross35 replied:
Looks good, thanks!

tgross35 commented:
I don't see an existing codegen test for the intrinsics so these should probably get one. https://github.com/rust-lang/rust/tree/3f33b30e19b7597a3acbca19e46d9e308865a0fe/tests/codegen/float would be a reasonable home.

calder commented on src/tools/miri/tests/pass/float.rs:
```diff
@@ -1144,6 +1144,46 @@ fn test_algebraic() {
+        assert_eq!(a.algebraic_div(b), a / b);
+        assert_eq!(a.algebraic_rem(b), a % b);
+    }
+
```
I figured it was a good idea to add these, but happy to remove or update the existing ones to use the stable API instead.

RalfJung replied:
We already have `test_algebraic` so this feels a bit redundant... also, the assertion will fail soon when https://github.com/rust-lang/miri/pull/4156 lands.

calder replied:
Reverted.

calder replied:
@RalfJung: Should we at least update these tests to use the public API? It feels weird to only test the internal API but not the public one.

RalfJung replied:
Ultimately the intrinsics are what Miri needs to implement. The doctests in this repo (most of them, anyway) are run in Miri as well so we should notice if doctests start using a different intrinsic we do not support yet.

calder commented on tests/codegen/float/f128-algebraic.rs:
```diff
@@ -0,0 +1,40 @@
+// CHECK-LABEL: fp128 @f128_algebraic_add(
+#[no_mangle]
+pub fn f128_algebraic_add(a: f128, b: f128) -> f128 {
+    // CHECK: fadd reassoc nsz arcp contract fp128 {{(%a, %b)|(%b, %a)}}
```
The addition and multiplication cases both end up as `%b, %a` rather than `%a, %b` which surprised me but isn't incorrect. I opted to allow either in case behavior changes in the future.

tgross35 commented on tests/codegen/float/f128-algebraic.rs:
```diff
@@ -0,0 +1,40 @@
+// Verify that algebraic intrinsics generate the correct LLVM calls for f128
```
Could you actually remove the return type from the `CHECK-LABEL` annotations? `f16` and `f128` have a different ABI on different platforms that would need to be handled with something like https://github.com/rust-lang/rust/blob/10baf4465a241bedad114cd27288475febb96388/tests/codegen/float/f128.rs#L63-L66, but we don't need to add the complexity to verify that here.

Also it would be best to smoosh the four test files into a single `algebraic.rs` to save the testsuite a few compilations.

Tests look good though, thanks for adding them 👍

calder replied:
Thanks! Done. Is it OK to keep the type in the internal `CHECK` lines since those aren't part of the ABI?

tgross35 replied:
Yeah that is good, it's just the function signatures that can change. Thanks!

calder commented:
Re adding examples to all the public functions: Did you have specific examples in mind? I'm struggling to think of ones that aren't repetitive / low signal-to-noise (i.e. assert this algebraic add is approximately equal to a normal add). Should we instead link to the central documentation and test approximate equality of each of the ops in a standalone test so it doesn't clutter the documentation?

Added central example with a couple small edits for brevity:

> For example:
>
> ```
> x = x.algebraic_add(x, a.algebraic_mul(b));
> ```
>
> May be rewritten as either:
>
> ```
> x = x + (a * b); // As written
> x = (a * b) + x; // Reordered to allow using a single `fma`
> ```

How's that look?

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13159938521/job/36725866247) [(plain)](https://github.com/rust-lang/rust/commit/230032b702fc6e0bb8f25974e1504e28d3aa76a2/checks/36725866247/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- library/core/src/primitive_docs.rs - prim_f32 (line 1336) stdout ----
error[E0425]: cannot find value `x` in this scope
 --> library/core/src/primitive_docs.rs:1337:1
  |
4 | x = x.algebraic_add(x, a.algebraic_mul(b));
  |
help: you might have meant to introduce a new binding
  |
  |
4 | let x = x.algebraic_add(x, a.algebraic_mul(b));
  | +++
error[E0425]: cannot find value `x` in this scope
 --> library/core/src/primitive_docs.rs:1337:5
  |
  |
4 | x = x.algebraic_add(x, a.algebraic_mul(b));

error[E0425]: cannot find value `x` in this scope
 --> library/core/src/primitive_docs.rs:1337:21
  |
  |
4 | x = x.algebraic_add(x, a.algebraic_mul(b));

error[E0425]: cannot find value `a` in this scope
 --> library/core/src/primitive_docs.rs:1337:24
  |
  |
4 | x = x.algebraic_add(x, a.algebraic_mul(b));

error[E0425]: cannot find value `b` in this scope
 --> library/core/src/primitive_docs.rs:1337:40
  |
  |
4 | x = x.algebraic_add(x, a.algebraic_mul(b));

error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0425`.
For more information about this error, try `rustc --explain E0425`.
Couldn't compile the test.
---- library/core/src/primitive_docs.rs - prim_f32 (line 1342) stdout ----
error[E0425]: cannot find value `x` in this scope
 --> library/core/src/primitive_docs.rs:1343:1
  |
4 | x = x + (a * b); // As written
  |
help: you might have meant to introduce a new binding
  |
  |
4 | let x = x + (a * b); // As written
  | +++
error[E0425]: cannot find value `x` in this scope
 --> library/core/src/primitive_docs.rs:1343:5
  |
  |
4 | x = x + (a * b); // As written

error[E0425]: cannot find value `a` in this scope
 --> library/core/src/primitive_docs.rs:1343:10
  |
  |
4 | x = x + (a * b); // As written

error[E0425]: cannot find value `b` in this scope
 --> library/core/src/primitive_docs.rs:1343:14
  |
  |
4 | x = x + (a * b); // As written

error[E0425]: cannot find value `x` in this scope
 --> library/core/src/primitive_docs.rs:1344:1
  |
  |
5 | x = (a * b) + x; // Reordered to allow using a single `fma`
  |
help: you might have meant to introduce a new binding
  |
  |
5 | let x = (a * b) + x; // Reordered to allow using a single `fma`
  | +++
error[E0425]: cannot find value `a` in this scope
 --> library/core/src/primitive_docs.rs:1344:6
  |
  |
5 | x = (a * b) + x; // Reordered to allow using a single `fma`

error[E0425]: cannot find value `b` in this scope
 --> library/core/src/primitive_docs.rs:1344:10
  |
  |
5 | x = (a * b) + x; // Reordered to allow using a single `fma`

error[E0425]: cannot find value `x` in this scope
 --> library/core/src/primitive_docs.rs:1344:15
  |
  |
5 | x = (a * b) + x; // Reordered to allow using a single `fma`

error: aborting due to 8 previous errors

For more information about this error, try `rustc --explain E0425`.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13160833529/job/36728883450) [(plain)](https://github.com/rust-lang/rust/commit/f06173708ffadbc368495ab7f4c033c8990249c9/checks/36728883450/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  |
4 | let x = 12.0;
  |     - you must specify a type for this binding, like `f32`
...
9 | let x = x.algebraic_add(x, a.algebraic_mul(b));


error[E0689]: can't call method `algebraic_mul` on ambiguous numeric type `{float}`
  |
5 | let a = 34.0;
  |     - you must specify a type for this binding, like `f32`
...
...
9 | let x = x.algebraic_add(x, a.algebraic_mul(b));

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0689`.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13161603527/job/36731524127) [(plain)](https://github.com/rust-lang/rust/commit/9cc7ba08100ff53a8490c57e41f743163e8921b8/checks/36731524127/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- library/core/src/primitive_docs.rs - prim_f32 (line 1336) stdout ----
error[E0658]: use of unstable library feature `float_algebraic`
 --> library/core/src/primitive_docs.rs:1340:7
  |
7 | x = x.algebraic_add(x, a.algebraic_mul(b));
  |
  = note: see issue #136469 <https://github.com/rust-lang/rust/issues/136469> for more information
  = note: see issue #136469 <https://github.com/rust-lang/rust/issues/136469> for more information
  = help: add `#![feature(float_algebraic)]` to the crate attributes to enable

error[E0658]: use of unstable library feature `float_algebraic`
 --> library/core/src/primitive_docs.rs:1340:26
  |
  |
7 | x = x.algebraic_add(x, a.algebraic_mul(b));
  |
  = note: see issue #136469 <https://github.com/rust-lang/rust/issues/136469> for more information
  = note: see issue #136469 <https://github.com/rust-lang/rust/issues/136469> for more information
  = help: add `#![feature(float_algebraic)]` to the crate attributes to enable

error[E0061]: this method takes 1 argument but 2 arguments were supplied
    --> library/core/src/primitive_docs.rs:1340:7
     |
     |
7    | x = x.algebraic_add(x, a.algebraic_mul(b));
     |       ^^^^^^^^^^^^^    ------------------ unexpected argument #2 of type `f32`
note: method defined here
    --> /checkout/library/core/src/num/f32.rs:1516:12
     |
     |
1516 |     pub fn algebraic_add(self, rhs: f32) -> f32 {
help: remove the extra argument
     |
     |
7    - x = x.algebraic_add(x, a.algebraic_mul(b));
7    + x = x.algebraic_add(x);

error: aborting due to 3 previous errors

Some errors have detailed explanations: E0061, E0658.
Some errors have detailed explanations: E0061, E0658.
For more information about an error, try `rustc --explain E0061`.
Couldn't compile the test.
---- library/core/src/primitive_docs.rs - prim_f32 (line 1345) stdout ----
error: value assigned to `x` is never read
 --> library/core/src/primitive_docs.rs:1350:1
  |
8 | x = (a * b) + x; // Reordered to allow using a single `fma`
  |
  = help: maybe it is overwritten before being read?
note: the lint level is defined here
 --> library/core/src/primitive_docs.rs:1343:9
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13162645576/job/36735089456) [(plain)](https://github.com/rust-lang/rust/commit/425963528ad087ab5ae0b9ce515af621f0513fc1/checks/36735089456/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- library/core/src/primitive_docs.rs - prim_f32 (line 1336) stdout ----
error[E0061]: this method takes 1 argument but 2 arguments were supplied
    --> library/core/src/primitive_docs.rs:1342:7
     |
9    | x = x.algebraic_add(x, a.algebraic_mul(b));
     |       ^^^^^^^^^^^^^    ------------------ unexpected argument #2 of type `f32`
note: method defined here
    --> /checkout/library/core/src/num/f32.rs:1516:12
     |
     |
1516 |     pub fn algebraic_add(self, rhs: f32) -> f32 {
help: remove the extra argument
     |
     |
9    - x = x.algebraic_add(x, a.algebraic_mul(b));
9    + x = x.algebraic_add(x);

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0061`.
```

</details>

calder commented:
Squashed and rebased around merge conflict. Only manual resolution needed:
```diff
diff --git a/library/std/tests/floats/lib.rs b/library/std/tests/floats/lib.rs
index ad82f1a44e7..0d475160778 100644
--- a/library/std/tests/floats/lib.rs
+++ b/library/std/tests/floats/lib.rs
@@ -1,4 +1,4 @@
-#![feature(f16, f128, float_gamma, float_minimum_maximum)] 
+#![feature(f16, f128, float_algebraic, float_gamma, float_minimum_maximum)]
  
 use std::fmt;
 use std::ops::{Add, Div, Mul, Rem, Sub};
```

tgross35 reviewed (approved):
This all looks good to me. We could probably enhance the docs more, but that can come later.

@RalfJung do you have anything?

calder commented:
@RalfJung: Any feedback or is it OK if we get this merged?

RalfJung commented:
Somehow seems like my reply-by-mail got lost:

Hey, have a bit of patience. I do this in my free time, and some other PRs have been waiting for my review a lot longer than 24h.

RalfJung commented on library/core/src/primitive_docs.rs:
```diff
@@ -1317,6 +1317,42 @@ mod prim_f16 {}
+///
+/// # Algebraic operators
+///
+/// Algebraic operators of the form `a.algebraic_*(b)` allow the compiler to reassociate individual
+/// floating point operations for better vectorization. The exact set of optimizations is
```
The first sentence says "reassociate" which points to a fairly specific transformation. Only later does it become clear that this is more general. I think we should word the first sentence so that if someone stops reading after that they don't have a completely wrong impression.
```suggestion
/// Algebraic operators of the form `a.algebraic_*(b)` allow the compiler to optimize
/// floating point operations using all the usual algebraic properties of real numbers --
/// despite the fact that those properties do *not* hold on floating point numbers.
/// This can give a great performance boost since it may unlock vectorization.
/// The exact set of optimizations is
```

RalfJung reviewed (commented):
Just one editorial comment.

calder commented:
Applied, reformatted, squashed and rebased. Thanks again for the review and sorry for the overeager ping!

jedbrown commented on library/core/src/primitive_docs.rs:
```diff
@@ -1317,6 +1317,45 @@ mod prim_f16 {}
+/// # let mut x: f32 = 12.0;
+/// # let a: f32 = 34.0;
+/// # let b: f32 = 56.0;
+/// x = x + (a * b); // As written
+/// x = (a * b) + x; // Reordered to allow using a single `fma`
```
I think this example is misleading since fp addition (and multiplication) are commutative (the rewriting is not significant) and it's the `AllowContract` flag that is necessary to allow use of fmuladd (when the target feature is available).

An alternate example would be `a + b + c + d` being rewritten as `(a + c) + (b + d)`, which has a shorter critical path and allows packed vector instruction to do `[a, b] + [c, d]`. If you prefer to illustrate with `AllowContract`, one could use `a * d - b * c` to be rewritten as `fmuladd(a, d, - b*c)`, which is only two instructions (but will produce nonzero if a=b, c=d).

calder replied:
Thanks! Updated, how's this look?

RalfJung commented:
Please avoid rebases when there was no conflict; Github is pretty bad at tracking changes to a PR in a force-push, making it nearly impossible to see what you changed since the last review.

RalfJung commented on library/core/src/primitive_docs.rs:
```diff
@@ -1317,6 +1317,50 @@ mod prim_f16 {}
+///
+/// The exact set of optimizations is unspecified but typically allows combining operations,
+/// rearranging series of operations based on mathematical properties, converting between division
+/// and reciprocal multiplication, and disregarding the sign of zero. This means that the results of
+/// elementary operations may have undefined precision, but will not invoke input-dependent
+/// undefined behavior.
```
```suggestion
/// and reciprocal multiplication, and disregarding the sign of zero. This means that the results of
/// elementary operations may have undefined precision, and "non-mathematical" values
/// such as NaN, +/- Inf, or -0.0 may behave in unexpected ways, but these operations
/// will never cause undefined behavior.
```

calder commented:
OK to squash and rebase for merge?

tgross35 commented:
I think we can get this in, please do squash 👍

calder commented:
Squashed and rebased.

tgross35 commented:
@bors r+

bors commented:
:pushpin: Commit d4c96795fd49557906d604517c95ca91e3d4262a has been approved by `tgross35`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=tgross35 d4c96795fd49557906d604517c95ca91e3d4262a -->
<!-- homu: {"type":"Approved","sha":"d4c96795fd49557906d604517c95ca91e3d4262a","approver":"tgross35","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Zalathar commented:
Failed in rollup: https://github.com/rust-lang/rust/pull/137199#issuecomment-2664577728

@bors r-

tgross35 commented on tests/codegen/float/algebraic.rs:
The methods probably aren't getting inlined, this should get fixed by adding:

```suggestion
// Ensure 
//@ compile-flags: -Copt-level=1
```

tgross35 commented on tests/codegen/float/algebraic.rs:
```diff
@@ -0,0 +1,146 @@
+// CHECK-LABEL: @f16_algebraic_add
+#[no_mangle]
+pub fn f16_algebraic_add(a: f16, b: f16) -> f16 {
+    // CHECK: fadd reassoc nsz arcp contract half {{(%a, %b)|(%b, %a)}}
```
Rust and LLVM are allowed to have intermediate ops or for whatever reason `fadd` something other than `%a` and `%b` directly. To avoid future noise, could you change all the tests to just match any vars rather than exact? The important part is just that we are adding the modifiers.

```suggestion
    // CHECK: fadd reassoc nsz arcp contract half %{{.*}}, %{{.*}}
```

tgross35 commented:
@bors try

(needs a squash before merge)

bors commented:
:hourglass: Trying commit 806af98e9c4b17b88ed215eb0da4003e9e248c99 with merge 621c0ed6002f4e2552c76981a8643dc0c058492d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"806af98e9c4b17b88ed215eb0da4003e9e248c99","merge_sha":"621c0ed6002f4e2552c76981a8643dc0c058492d"} -->

calder commented:
Squashing and rebasing.

calder committed:
Expose algebraic floating point intrinsics

calder commented:
@tgross35: Any more feedback or can we get this merged?

bors commented:
:hourglass: Trying commit 6e1ac39a4a19c9875eb873d2efbe04b4b83b37bc with merge e73487c5d1c83637bdfa263707c3fa330ff7fded...
<!-- homu: {"type":"TryBuildStarted","head_sha":"6e1ac39a4a19c9875eb873d2efbe04b4b83b37bc","merge_sha":"e73487c5d1c83637bdfa263707c3fa330ff7fded"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13535606907/job/37829581108)
Build commit: e73487c5d1c83637bdfa263707c3fa330ff7fded (`e73487c5d1c83637bdfa263707c3fa330ff7fded`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13535606907/job/37829581108"},"merge_sha":"e73487c5d1c83637bdfa263707c3fa330ff7fded"} -->

tgross35 commented:
@bors r+

bors commented:
:pushpin: Commit 6e1ac39a4a19c9875eb873d2efbe04b4b83b37bc has been approved by `tgross35`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=tgross35 6e1ac39a4a19c9875eb873d2efbe04b4b83b37bc -->
<!-- homu: {"type":"Approved","sha":"6e1ac39a4a19c9875eb873d2efbe04b4b83b37bc","approver":"tgross35","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

fmease commented:
@bors try

bors commented:
:hourglass: Trying commit 6e1ac39a4a19c9875eb873d2efbe04b4b83b37bc with merge cd32041858d8d59256dfb82addd9e81ae09cccd5...
<!-- homu: {"type":"TryBuildStarted","head_sha":"6e1ac39a4a19c9875eb873d2efbe04b4b83b37bc","merge_sha":"cd32041858d8d59256dfb82addd9e81ae09cccd5"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-aux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13550410511/job/37872445393) [(plain)](https://github.com/rust-lang-ci/rust/commit/cd32041858d8d59256dfb82addd9e81ae09cccd5/checks/37872445393/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
failures:

---- f16::test_algebraic stdout ----

thread 'f16::test_algebraic' panicked at library/std/tests/floats/f16.rs:964:5:
0x607e is not approximately equal to 0x6086 (threshold 1e-6, difference 0x4400)
---- f32::test_algebraic stdout ----


thread 'f32::test_algebraic' panicked at library/std/tests/floats/f32.rs:924:5:
578.9989 is not approximately equal to 579.0 (threshold 1e-6, difference 0.0010986328)

failures:
    f16::test_algebraic
    f32::test_algebraic
    f32::test_algebraic

test result: FAILED. 152 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 29.81s

error: test failed, to rerun pass `-p std --test floats`
Build completed unsuccessfully in 0:05:46
make: *** [Makefile:57: check-aux] Error 1
  network time: Wed, 26 Feb 2025 19:36:02 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13550410511/job/37877443453)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/13550410511/job/37877443453","builder_name":"checks-actions"} -->

pthariensflame commented on library/core/src/primitive_docs.rs:
````diff
@@ -1317,6 +1317,51 @@ mod prim_f16 {}
+/// x = a.algebraic_add(b).algebraic_add(c).algebraic_add(d);
+/// ```
+///
+/// May be rewritten as either:
````
```suggestion
/// May be rewritten as:
```

There’s only one example that follows!

calder replied:
Nope, 2:
```rust
x = a + b + c + d; // As written
x = (a + c) + (b + d); // Reordered to shorten critical path and enable vectorization
```

tgross35 commented:
@calder would you be able to update this? The thresholds likely need to be increased

calder commented:
@tgross35: Fixed tolerances, sorry for the delay!

tgross35 commented:
@bors try

It looks like this PR includes an accidental backtrace downgrade, not sure why rustbot didn't flag that but it should be dropped.

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13782819007/job/38549017498)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/13782819007/job/38549017498","builder_name":"checks-actions"} -->

calder commented:
@bors try

Tweaked tolerances some more. Is there an easy way to run these MIRI tests locally?

bors commented:
@calder: :key: Insufficient privileges: not in try users

bors commented:
:hourglass: Trying commit 0151a019987a15d40700439b903bc51ac1446b73 with merge dabe80ae85417b4bedb48d495fdec82e418e590c...
<!-- homu: {"type":"TryBuildStarted","head_sha":"0151a019987a15d40700439b903bc51ac1446b73","merge_sha":"dabe80ae85417b4bedb48d495fdec82e418e590c"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-aux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13797721528/job/38593316138) [(plain)](https://github.com/rust-lang-ci/rust/commit/dabe80ae85417b4bedb48d495fdec82e418e590c/checks/38593316138/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.url=https://github.com/rust-lang/rust
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
failures:

---- f32::test_algebraic stdout ----

thread 'f32::test_algebraic' panicked at library/std/tests/floats/f32.rs:924:5:
579.0011 is not approximately equal to 579.0 (threshold 0.001, difference 0.0010986328)


failures:
    f32::test_algebraic

test result: FAILED. 153 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 30.07s

error: test failed, to rerun pass `-p std --test floats`
Build completed unsuccessfully in 0:06:13
make: *** [Makefile:57: check-aux] Error 1
  local time: Tue Mar 11 22:15:48 UTC 2025
  network time: Tue, 11 Mar 2025 22:15:48 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13797721528/job/38598155593)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/13797721528/job/38598155593","builder_name":"checks-actions"} -->

tgross35 commented:
@bors delegate+ for `@bors try`, just don't merge it (I'll take a final look once things pass)

tgross35 commented:
@bors try

It's a bit better to use `--keep-base` when you rebase locally to squash, that way GH shows a diff of only relevant things that changed. Unfortunately GH has the limitation that if the push includes an actual rebase onto latest (sometimes unavoidable ofc), then the "compare" link is pretty unhelpful e.g. https://github.com/rust-lang/rust/compare/53944e53dca5f93f1559b3914470c61dda2c350c..dfd4b6cddf5b2a3936492fe5c7cb97198fec3761

bors commented:
:hourglass: Trying commit dfd4b6cddf5b2a3936492fe5c7cb97198fec3761 with merge 2d8ed6b4bf07905ce9e8b131b366708f273717c0...
<!-- homu: {"type":"TryBuildStarted","head_sha":"dfd4b6cddf5b2a3936492fe5c7cb97198fec3761","merge_sha":"2d8ed6b4bf07905ce9e8b131b366708f273717c0"} -->

tgross35 commented on library/std/tests/floats/f32.rs:
```diff
@@ -915,3 +915,15 @@ fn test_total_cmp() {
     assert_eq!(Ordering::Less, (-s_nan()).total_cmp(&s_nan()));
 }
+
+#[test]
+fn test_algebraic() {
+    let a: f32 = 123.0;
+    let b: f32 = 456.0;
+
+    assert_approx_eq!(a.algebraic_add(b), a + b, 1e-2);
+    assert_approx_eq!(a.algebraic_sub(b), a - b, 1e-2);
+    assert_approx_eq!(a.algebraic_mul(b), a * b, 1e-1);
+    assert_approx_eq!(a.algebraic_div(b), a / b, 1e-5);
+    assert_approx_eq!(a.algebraic_rem(b), a % b, 1e-2);
+}
```
What is actually going on that causes these tests to fail? `a + b` should be 579.0, but the result is 579.0011. Afaict none of the algebraic effects can come into play here, so shouldn't the result be exact?

calder replied:
@LorrensP-2158466 @RalfJung can confirm but I assume this is from https://github.com/rust-lang/miri/commit/f5330d02364a2bfd987d2f356ae399c1a4a59039 since this works under `./x test` but fails under `./x miri`:
```rust
assert_eq!(a.algebraic_add(b), 579.0);
assert_eq!(b.algebraic_add(a), 579.0);
```

Would it make sense to specialize this test for MIRI vs. non-MIRI and assert exact values in the non-MIRI variant?

calder replied:
Here's what that would look like: 6ab3fec16ab63c86990dbc0186392e065fd65038

LorrensP-2158466 replied:
Typing on my phone, so I'm sorry in advanced:).

In a PR a while ago I introduced some extra non-determinism on floating point operations with no specified precision. Sometime later Ralf disabled these because they failed tests in std lib and others. But I think not the algebraic ones. These get a 16ULP error on top of the host error.

There is new PR where we are  working on lowering these errors and accounting for this in the rust tests.

I think the best thing to do is just disabling this ULP error everywhere and take this on in the current pr or a follow up.

But Ralf knows best ofc ;)

calder replied:
Thanks! Re:
> I think the best thing to do is just disabling this ULP error everywhere and take this on in the current pr or a follow up.

Do you mean revert 622e8f4812b8bcd9a3501ca7d299987fff733b68 in its entirety, or temporarily make `apply_random_float_error()` a noop?

LorrensP-2158466 replied:
I would suggest doing what Ralf said in his last comment, and if that still gives problems, you can do this in `src/tools/miri/src/intrinsics/mod.rs`:
```rust
// FIXME: temporarily disabled as it breaks std tests.
// let res = apply_random_float_error_ulp(
//     this,
//     res,
//     4, // log2(16)
// );
```

This is already done for `sin`, `cos`, `pow`, ... but not for the others, so I guess that would be fine.

LorrensP-2158466 replied:
But if everybody is okey with the no-op route that's also good way to temporarily fix this.

RalfJung replied:
We partially reverted the Miri change to be able to properly deal with the existing tests, but I don't think there's any reason to do the same for new tests. We can tweak the error to 4 ULP maybe, but the fact that an error is applied should remain. Tests that rely on non-guarantees properties are questionable so I'd like to know why you want to test for full equality here in the first place -- this test can legitimately fail without Miri.

calder replied:
OK. I'm good with reverting to this to keep things simple:
```rust
 #[test]
 fn test_algebraic() {
     let a: f32 = 123.0;
     let b: f32 = 456.0;
 
     assert_approx_eq!(a.algebraic_add(b), a + b, 1e-2);
     assert_approx_eq!(a.algebraic_sub(b), a - b, 1e-2);
     assert_approx_eq!(a.algebraic_mul(b), a * b, 1e-1);
     assert_approx_eq!(a.algebraic_div(b), a / b, 1e-5);
     assert_approx_eq!(a.algebraic_rem(b), a % b, 1e-2);
 }
```

@tgross35: You OK with that?

@RalfJung: Is there a better way to define these error terms?

RalfJung replied:
Not sure what you are asking about the error terms?

I'd say we can reduce the error to 4 ULP, then the error shouldn't be as big any more. But on f32, even just 4 ULP can already be a non-trivial absolute error.

This is controlled here:

https://github.com/rust-lang/rust/blob/716dd22c04e936853b9bc963da616e1b28525e6b/src/tools/miri/src/intrinsics/mod.rs#L416

And then these should probably have the same error:

https://github.com/rust-lang/rust/blob/716dd22c04e936853b9bc963da616e1b28525e6b/src/tools/miri/src/intrinsics/mod.rs#L469

calder replied:
Updated. How's 8ac2e73180d412434f05dd107152c4b3b1cfa3b8 look? Should we also update these for consistency or leave them as is?
* https://github.com/rust-lang/rust/blob/716dd22c04e936853b9bc963da616e1b28525e6b/src/tools/miri/src/intrinsics/mod.rs#L252
* https://github.com/rust-lang/rust/blob/716dd22c04e936853b9bc963da616e1b28525e6b/src/tools/miri/src/intrinsics/mod.rs#L288
* https://github.com/rust-lang/rust/blob/716dd22c04e936853b9bc963da616e1b28525e6b/src/tools/miri/src/intrinsics/mod.rs#L503

> Not sure what you are asking about the error terms?

Sorry, I was asking if there's a constant or helper I should use instead of hardcoding absolute error terms in these `assert_approx_eq()` calls. I don't see any that make sense though.

calder replied:
Copy, thanks! All good as is then. I'll squash so we can get this merged.

tgross35 replied:
> this test can legitimately fail without Miri.

Do you mean within the limits of what we document, or in practice with what we currently have? I don't think any of `reassoc nsz arcp contract` allow imprecise results for only a single float op, with the possible exception of `arcp`.

RalfJung replied:
Fair, I should have been more clear -- I mean within the limits of the spec.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13820612362/job/38670404123)
Build commit: 2d8ed6b4bf07905ce9e8b131b366708f273717c0 (`2d8ed6b4bf07905ce9e8b131b366708f273717c0`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13820612362/job/38670404123"},"merge_sha":"2d8ed6b4bf07905ce9e8b131b366708f273717c0"} -->

leonardo-m commented:
By they way, im my codebase I use something minimal like this:

```rust
#[derive(Copy, Clone, Default)]
struct Af64(pub f64);

impl Af64 {
    fn from(x: i32) -> Self { Self(f64::from(x)) }
}

impl Add for Af64 {
    type Output = Self;

    #[inline]
    fn add(self, other: Self) -> Self {
        Self(fadd_algebraic(self.0, other.0))
    }
}

impl Sub for Af64 {
    type Output = Self;

    #[inline]
    fn sub(self, other: Self) -> Self {
        Self(fsub_algebraic(self.0, other.0))
    }
}

impl Mul for Af64 {
    type Output = Self;

    #[inline]
    fn mul(self, other: Self) -> Self {
        Self(fmul_algebraic(self.0, other.0))
    }
}

impl Div for Af64 {
    type Output = Self;

    #[inline]
    fn div(self, other: Self) -> Self {
        Self(fdiv_algebraic(self.0, other.0))
    }
}

impl Rem for Af64 {
    type Output = Self;

    #[inline]
    fn rem(self, other: Self) -> Self {
        Self(frem_algebraic(self.0, other.0))
    }
}
```

calder commented:
Squashed with `--keep-base` and rebased past merge conflict but accidentally pushed intermediate, sorry!

RalfJung commented:
We explicitly do not guarantee an exact value so it seems odd to have a test that checks for an exact value...

tgross35 commented:
@bors try

bors commented:
:hourglass: Trying commit 99a96f47174524942ccb8519c784a6da75e678fd with merge 14c0dd4827575e8311606bd34e428ae57e5cf472...
<!-- homu: {"type":"TryBuildStarted","head_sha":"99a96f47174524942ccb8519c784a6da75e678fd","merge_sha":"14c0dd4827575e8311606bd34e428ae57e5cf472"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-aux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/14227519446/job/39870541051) [(plain)](https://github.com/rust-lang-ci/rust/commit/14c0dd4827575e8311606bd34e428ae57e5cf472/checks/39870541051/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.url=https://github.com/rust-lang/rust
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
failures:

---- f16::test_algebraic stdout ----

thread 'f16::test_algebraic' panicked at library/std/tests/floats/f16.rs:964:5:
0x7ade is not approximately equal to 0x7ad9 (threshold 100.0, difference 0x5900)


failures:
    f16::test_algebraic

test result: FAILED. 154 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 28.62s

error: test failed, to rerun pass `-p std --test floats`
Build completed unsuccessfully in 0:05:37
make: *** [Makefile:57: check-aux] Error 1
  local time: Wed Apr  2 20:40:48 UTC 2025
  network time: Wed, 02 Apr 2025 20:40:48 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14227519446/job/39875819906)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/14227519446/job/39875819906","builder_name":"checks-actions"} -->

RalfJung commented:
If you prefer you can remove the non-determinism from the algebraic operations in Miri and we'll figure out the best way to deal with that in a follow-up PR. We should figure this out before stabilization to ensure we have a coherent semantics, but it doesn't have to all be fully done in the first PR.

calder commented:
@tgross35: One more try? (Let me know if there's a way to run all these tests locally to speed up the iteration process)

RalfJung commented:
@bors try

You can run this locally with `./x miri std -- algebraic`.

bors commented:
:hourglass: Trying commit 9602bbbebdadbcc00b44d8805b75615d228f9c1b with merge 7a97f5ea73db9a162b151c4dda7cde3c8fc1afc0...
<!-- homu: {"type":"TryBuildStarted","head_sha":"9602bbbebdadbcc00b44d8805b75615d228f9c1b","merge_sha":"7a97f5ea73db9a162b151c4dda7cde3c8fc1afc0"} -->

calder commented:
Yeah, that's what I've been running but the last failure was on `x86_64-gnu-aux` not Miri. Is there a good way to run the full suite that Bors is running?

RalfJung commented:
> the last failure was on x86_64-gnu-aux not Miri.

x86_64-gnu-aux is the name of a test runner. That runner runs many tests. Among them, it runs `./x miri`.

calder commented:
Gotcha, thanks! Confirmed I can reproduce with:
```sh
MIRIFLAGS="-Zmiri-many-seeds" ./x miri std -- algebraic
```

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14246541490/job/39936150889)
Build commit: 7a97f5ea73db9a162b151c4dda7cde3c8fc1afc0 (`7a97f5ea73db9a162b151c4dda7cde3c8fc1afc0`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14246541490/job/39936150889"},"merge_sha":"7a97f5ea73db9a162b151c4dda7cde3c8fc1afc0"} -->

tgross35 commented:
Could the tests get something like `let approx = if cfg!(miri) { 1e-3 } else { 0.0 };` and use that for the `assert_approx_eq!` tolerance? Similar to https://github.com/rust-lang/rust/pull/138062 the API doesn't guarantee much so it makes sense to have the imprecision in Miri. This current implementation _should_ be exact though so it would be nice if the internal tests flag any problems there.

RalfJung commented:
Such logic should be accompanied by comments explaining that this test does *not* imply any form of guarantee, it is merely a sentinel for us to notice when the real precision of these operations goes down.

calder commented:
Updated all tests to assert exact equality outside of MIRI. To support this cleanly I update `assert_approx_eq()` to assert  `diff <= epsilon` rather than `diff < epsilon`: https://github.com/rust-lang/rust/pull/136457/commits/266ca8b23573d6719aa3b7b07c3bb6aa706861f6#diff-a4ab8b2578f382b20890fc918f6cdbffcaa057078f8793f0d974c011003a2b01L13-R13

If we don't want to loosen that we'll need to do something like this instead:
```diff
diff --git a/library/std/tests/floats/f32.rs b/library/std/tests/floats/f32.rs
index d99b03cb255..c75e652dc1f 100644
--- a/library/std/tests/floats/f32.rs
+++ b/library/std/tests/floats/f32.rs
@@ -926,13 +926,17 @@ fn test_algebraic() {
     // This is a check of current implementations and does NOT imply any form of
     // guarantee about future behavior. The compiler reserves the right to make
     // these operations inexact matches in the future.
-    let eps_add = if cfg!(miri) { 1e-3 } else { 0.0 };
-    let eps_mul = if cfg!(miri) { 1e-1 } else { 0.0 };
-    let eps_div = if cfg!(miri) { 1e-4 } else { 0.0 };
-
-    assert_approx_eq!(a.algebraic_add(b), a + b, eps_add);
-    assert_approx_eq!(a.algebraic_sub(b), a - b, eps_add);
-    assert_approx_eq!(a.algebraic_mul(b), a * b, eps_mul);
-    assert_approx_eq!(a.algebraic_div(b), a / b, eps_div);
-    assert_approx_eq!(a.algebraic_rem(b), a % b, eps_div);
+    if cfg!(miri) {
+        assert_approx_eq!(a.algebraic_add(b), a + b, 1e-3);
+        assert_approx_eq!(a.algebraic_sub(b), a - b, 1e-3);
+        assert_approx_eq!(a.algebraic_mul(b), a * b, 1e-1);
+        assert_approx_eq!(a.algebraic_div(b), a / b, 1e-4);
+        assert_approx_eq!(a.algebraic_rem(b), a % b, 1e-4);
+    } else {
+        assert_eq!(a.algebraic_add(b), a + b);
+        assert_eq!(a.algebraic_sub(b), a - b);
+        assert_eq!(a.algebraic_mul(b), a * b);
+        assert_eq!(a.algebraic_div(b), a / b);
+        assert_eq!(a.algebraic_rem(b), a % b);
+    }
 }
```

tgross35 commented:
Latest changes lgtm. r=me with a squash and a passing try

@bors try

bors commented:
:hourglass: Testing commit 8ff70529f22158dab9a06c7b2e79e1d460d24c61 with merge 994f291efdb7969b06dff1ade574e447dbfacd91...
<!-- homu: {"type":"BuildStarted","head_sha":"8ff70529f22158dab9a06c7b2e79e1d460d24c61","merge_sha":"994f291efdb7969b06dff1ade574e447dbfacd91"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14275745775/job/40020986905)
Build commit: 994f291efdb7969b06dff1ade574e447dbfacd91 (`994f291efdb7969b06dff1ade574e447dbfacd91`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14275745775/job/40020986905"},"merge_sha":"994f291efdb7969b06dff1ade574e447dbfacd91"} -->

tgross35 commented:
@bors r+

bors commented:
:pushpin: Commit 8ff70529f22158dab9a06c7b2e79e1d460d24c61 has been approved by `tgross35`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=tgross35 8ff70529f22158dab9a06c7b2e79e1d460d24c61 -->
<!-- homu: {"type":"Approved","sha":"8ff70529f22158dab9a06c7b2e79e1d460d24c61","approver":"tgross35","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
